include/seqan3/core/concept/cereal.hpp

```diff
@@ -0,0 +1,153 @@
+//                 SeqAn - The Library for Sequence Analysis
+// ============================================================================
+//
+// Copyright (c) 2006-2017, Knut Reinert & Freie Universitaet Berlin
+// Copyright (c) 2016-2017, Knut Reinert & MPI Molekulare Genetik
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \brief Adaptions of concepts from the Cereal library.
+ * \ingroup core
+ * \author Hannes Hauswedell <hannes.hauswedell AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <type_traits>
+
+#include <seqan3/core/platform.hpp>
+
+#if SEQAN_WITH_CEREAL
+
+#include <cereal/details/traits.hpp>
+
+namespace seqan3
+{
+
+template <typename t>
+concept bool cereal_output_archive_concept = std::is_base_of_v<cereal::detail::OutputArchiveBase, t>;
+
+template <typename t>
+concept bool cereal_input_archive_concept = std::is_base_of_v<cereal::detail::InputArchiveBase, t>;
+
+template <typename t>
+concept bool cereal_archive_concept = cereal_output_archive_concept<t> || cereal_input_archive_concept<t>;
+
+template <typename t>
+concept bool cereal_text_archive_concept = std::is_base_of_v<cereal::traits::TextArchive, t>;
+
+} // namespace seqan3
+
+namespace seqan3::detail
+{
+
+template <typename type>
+using strip_cereal_wrapper_t = typename cereal::traits::strip_minimal<std::decay_t<type>>::type;
```

Comment:
The stuff in this file could all be done with a `static constexpr bool seqan_with_serial` that's no problem. But in other cases it's not possible.

Reply:
>so the second code path is likely not to be as tested as the other one.
> Thus it might happen that we add something to the first code path and will forgot to update the second one.

It definitely needs the same amount of testing.

> You could also make local #ifdefs like

Yeah, I could re-order it, but it increases the macro verbosity.